design/0001-frr.md

```diff
@@ -0,0 +1,411 @@
+
+The container will expose a control interface that will initially be used to
+configure FRR via the `vtysh` application and eventually via its gRPC golang
+bindings. This will be used to implement bgp.New() and bgp.Close() for session
+creation and bgp.Set() for prefix advertisement.
+
+#### others
+
+There may be some common functionality in the bgp package that could be reused
+between BGP implementations. For example, the new bgp package will require
+integration with the “metrics” struct which updates prometheus in response to
+BGP events. It may be necessary to refactor these types of functionality into a
+separate package for reuse amongst BGP implementations.
+
+### Test Plan
+
+**e2e tests**: The intention is to expand on the MetalLB ‘dev-env’ KIND
+environment. After deploying this environment, it will be possible to run any
+relevant Kubernetes e2e tests (or custom MetalLB tests) against this test
+cluster. Further investigation will be required as part of [Story 1](#story-1)
+in order to determine the relevance of any Kubernetes e2e tests and if any other
+e2e tests will need to be developed.
+**unit tests**: Unit tests will be added to
+any additional code in the MetalLB repository.
+**upgrade**: We may need to add
+tests to deal with an upgrade to a newer FRR version.
+
+* We will allow independent upgrade of the FRR component in order to resolve any
+  specific FRR bugs without upgrading MetalLB.
+* This enhancement assumes that each node’s MetalLB ‘speaker’ component can be
+  stopped (for an arbitrary time) and restarted without disruption of MetalLB as
+  a whole (To be confirmed)
+* This enhancement assumes that FRR will be deployed as a DaemonSet on the
+  Kubernetes cluster. It is also assumed that there is a stable/versioned
+  interface between this DaemonSet and MetalLB. Therefore, if any non-backwards
+  compatible changes are made in this interface, this will need to be managed in
+  the MetalLB ‘speaker’ component.
+* Upgrade/downgrade will remove a node from the MetalLB memberlist by unlabeling
+  the node. At this point the ‘speaker’ component on the node can be stopped and
+  the FRR component can be updated to the new version. It should be noted that
+  as FRR only peers with BGP agents outside the cluster, there is no requirement
+  that each node within the cluster maintains the same version or a compatible
+  version of FRR. However, it is required that each version of FRR is compatible
+  with the BGP agent with which it is peering. After FRR is updated, the
+  ‘speaker’ component can be restarted and the node can be labelled to add it to
+  the memberlist again.
+
+### Version Skew Strategy
+
+Version skew between versions of FRR should not be a concern to Kubernetes or
+MetalLB as long as FRR presents a stable interface to the MetalLB ‘speaker’
+component. This is because FRR will only peer with BGP agents outside of the
+cluster and not with each other.
+
+## Drawbacks
```

Comment:
Another drawback IMO is expanding the speaker's attack surface. In terms of security I'd say it's a plus to have a very thin BGP implementation. For example, you can't confuse MetalLB by sending malicious advertisements to the speaker because it doesn't support *accepting* BGP updates, only sending them.

We *may* be able to mitigate this using FRR config hardening, I'm not familiar with FRR to know for sure. Anyway, I'd say this is a point to consider.

Reply:
Good point, I will update and add some examples configurations that I can find.